### ISSUE TITLE
PrimeFaces: Incompatibility with Oracle Weblogic 14 & HTTPS on HTTP2

### ISSUE THREAD

**Issue by pascalbxlx** [primefaces/primefaces#7243]
We are porting an existing application from weblogic 12 to weblogic 14 using https and we are facing issues. We have reported this to Oracle and the feedback from them:

_The problem is introduced by HTTP/2. 

WLS 12c is using HTTP1.1 protocol, this is why the application is working in WLS 12c.
WLS 14c is using HTTP/2 protocol. If HTTP/2 is manually disabled in Chrome, you will see that it works ok.

The browser closes the socket and the server cannot write back anymore. The common socket wanted by HTTP/2 -handler is closed before all the resources are written out. 
PrimeFaces is using quite a few javascript and style resources and is stuck when failing to load these._

When trying to render the page we get some exceptions. Here is the explanation:

Analysis:
=========
As noted in the earlier analysis in December, the failure is due to Broken pipe and socket closed exceptions. 
This used to work sometimes in Firefox correctly after disabling the timeout. It also works in Chrome browser but the stylesheets were not getting loaded.
 
The server log has consistent failures related with Stream ID for payload : MuxableSocketHTTP2 Received : line 326[0, 0, 5, 2, 0, 0, 0, 0, 2, -128, 0, 0, 1, 109]>
 
####<Apr 14, 2021, 10:27:21,284 PM Pacific Daylight Time> <Debug> <Http>
<slc15baj> <myserver> <ExecuteThread: '2' for queue: 'weblogic.socket.Muxer'>
<<WLS Kernel>> <> <ea507828-3c3e-4ae4-8f1e-e66615d924fe-0000002e>
<1618464441284> <[severity-value: 128] [rid: 0] > <BEA-000000>
<MuxableSocketHTTP2 catching Exception:
weblogic.servlet.http2.ConnectionException: Failed to create remote stream,
the id [6] must be even
at
weblogic.servlet.http2.MuxableSocketHTTP2$StreamManagerImpl.createRemoteStream
(MuxableSocketHTTP2.java:762)
at
weblogic.servlet.http2.MuxableSocketHTTP2$StreamManagerImpl.createRemoteStream
(MuxableSocketHTTP2.java:649)
at
weblogic.servlet.http2.FrameHandler.findOrCreateStream(FrameHandler.java:403)
 
Prior to this stream, the browser sends a SETTING frame:
 
<MuxableSocketHTTP2 Received : line 326[0, 0, 0, 4, 1, 0, 0, 0, 0]>
Line 26793: ####<Apr 14, 2021, 10:27:19,320 PM Pacific Daylight Time>
<Debug> <Http> <slc15baj> <myserver> <ExecuteThread: '2' for queue:
'weblogic.socket.Muxer'> <<WLS Kernel>> <>
<ea507828-3c3e-4ae4-8f1e-e66615d924fe-0000002e> <1618464439320>
<[severity-value: 128] [rid: 0] > <BEA-000000> <Receiving frame: type=SETTINGS, flags=1, streamId=0 on Connection:
weblogic.servlet.http2.MuxableSocketHTTP2@701f4288:Socket[addr=/10.191.194.128,port=64980,localport=7002]>
 
https://tools.ietf.org/html/rfc7540#section-6.5
 
Diagnosis:
==========
The provided application is huge and identifying the component that sends the buffer which is interpreted as sometimes PRIORITY and sometimes RST_STREAM (both after the SETTINGS frame received) was a challenge, therefore the delay in providing the final conclusion.
The URI is always /taxud/uumds/admin-int.
The logs show that resources that are trying to get loaded are
/resources/components/datatable/dataTable.xhtml,
/taxud/uumds/admin-int/javax.faces.resource/css/dss.css.xhtml etc. 
 
All these resources have PrimeFaces dependencies.
 
To further demonstrate the PrimeFace root cause, Oracle development created two sample apps one with Primefaces and one without. 
The application which has PrimeFaces dependency is consistently failing with Broken Pipe/Socket closed exceptions. 
The simple JSF application without PrimeFaces is working fine.
You can find them attached to the SMT ticket IM0019281133 (the email server rejected sending them as attachments) and you can deploy and test each of them in the browser, while checking the logs in parallel, for each execution.
 
Please let me know your opinion on the analysis and the conclusion. 
In addition, if you would like to discuss with Oracle development directly, I can also setup a call.

**Comment by tandraschko** [comment]
i dont think this is PrimeFaces related.
ResourceHandling is done directly in JSF, JSF just streams PrimeFaces resources.

The "broken pipe" is also expected as the client aborts the requests. This is because of a Firefox features, discussed several times in the forum.

In MyFaces we reduced the logging in this case, see here: https://issues.apache.org/jira/browse/MYFACES-3736

**Comment by GedMarc** [comment]
Http2 is only supported from jakarta EE 8 / Servlet 4.1/5, which is weblogic 14.... 
You are attempting to do something that your server does not handle

Not a primefaces bug, client implementation error. On servers that fully support it, there is no problems
Undertow 2.2 etc,,,

**Comment by melloware** [comment]
@pascalbxlx You also didn't mention which versions of Mojarra that WL 12 and 14 use?     Is WebLogic pluggable where its possible to swap Mojarra out for MyFaces?  That is what we do with some of our Jboss installations for better performance.

**Comment by melloware** [comment]
@pascalbxlx It sounds like you have a very large payload causing your broken pipes as @tandraschko mentioned.  I am going to close this ticket for now as WontFix as there is nothing for PF to do about this.  If you do have more info or want us to reopen because you have found something definitive PF can fix I will reopen!

**Comment by pascalbxlx** [comment]
> Http2 is only supported from jakarta EE 8 / Servlet 4.1/5, which is weblogic 14....
> You are attempting to do something that your server does not handle
> 
> Not a primefaces bug, client implementation error. On servers that fully support it, there is no problems
> Undertow 2.2 etc,,,

But indeed. Here we are migrating from WL12 to WL14 and the problem appears on WL14.. The problem does not happen on WL12

**Comment by pascalbxlx** [comment]
> @pascalbxlx You also didn't mention which versions of Mojarra that WL 12 and 14 use? Is WebLogic pluggable where its possible to swap Mojarra out for MyFaces? That is what we do with some of our Jboss installations for better performance.

I will ask Oracle. We took the latest package available from Oracle. There is no pluggable mechanism but may be it could be overriding by putting a specific Mojarra in the patch path. In any case I need the support of Oracle (the installation has to be supported).

**Comment by melloware** [comment]
Yes please let us know what Oracle says and if we can help.
